text-serializer-moshi/src/main/java/net/kyori/adventure/text/serializer/moshi/StyleAdapter.java

```diff
@@ -0,0 +1,304 @@
+    }
+
+    for (int i = 0, length = DECORATIONS.length; i < length; i++) {
+      final TextDecoration decoration = DECORATIONS[i];
+      final String name = TextDecoration.NAMES.key(decoration);
+      if (json.containsKey(name)) {
+        style.decoration(decoration, asBoolean(json.get(name)));
+      }
+    }
+
+    if (json.containsKey(INSERTION)) {
+      style.insertion(asString(json.get(INSERTION)));
+    }
+
+    if (json.containsKey(CLICK_EVENT)) {
+      final Map<String, Object> clickEvent = (Map<String, Object>) json.get(CLICK_EVENT);
+      if (clickEvent != null) {
+        final ClickEvent.Action action = this.optionallyDeserialize(clickEvent.get(CLICK_EVENT_ACTION), ClickEvent.Action.class);
+        if (action != null && action.readable()) {
+          final @Nullable Object rawValue = clickEvent.get(CLICK_EVENT_VALUE);
+          final @Nullable String value = rawValue == null ? null : asString(rawValue);
+          if (value != null) {
+            style.clickEvent(ClickEvent.clickEvent(action, value));
+          }
+        }
+      }
+    }
+
+    if (json.containsKey(HOVER_EVENT)) {
+      final Map<String, Object> hoverEvent = (Map<String, Object>) json.get(HOVER_EVENT);
+      if (hoverEvent != null) {
+        final HoverEvent.Action action = this.optionallyDeserialize(hoverEvent.get(HOVER_EVENT_ACTION), HoverEvent.Action.class);
+        if (action != null && action.readable()) {
+          final @Nullable Object value;
+          if (hoverEvent.containsKey(HOVER_EVENT_CONTENTS)) {
+            final @Nullable Object rawValue = hoverEvent.get(HOVER_EVENT_CONTENTS);
+            value = this.moshi.adapter(action.type()).fromJsonValue(rawValue);
+          } else if (hoverEvent.containsKey(HOVER_EVENT_VALUE)) {
+            final Component rawValue = this.moshi.adapter(Component.class).fromJsonValue(hoverEvent.get(HOVER_EVENT_VALUE));
+            value = this.legacyHoverEventContents(action, rawValue);
+          } else {
+            value = null;
+          }
+
+          if (value != null) {
+            style.hoverEvent(HoverEvent.hoverEvent(action, value));
+          }
+        }
+      }
+    }
+
+    return style.build();
+  }
+
+  @Override
+  @SuppressWarnings({"rawtypes", "unchecked"}) // Screw you too Java generics
+  public void toJson(final JsonWriter writer, final Style value) throws IOException {
+    // This is a hack that is used to signal that the writer is writing a Component, and so we shouldn't begin/end an object.
+    final Component tag = writer.tag(Component.class);
+    if (tag == null) {
+      writer.beginObject();
+    }
+
+    for (int i = 0, length = DECORATIONS.length; i < length; i++) {
+      final TextDecoration decoration = DECORATIONS[i];
+      final TextDecoration.State state = value.decoration(decoration);
+      if (state != TextDecoration.State.NOT_SET) {
+        final String name = TextDecoration.NAMES.key(decoration);
+        assert name != null; // should never be null
```

Comment:
Why not `decoration.toString()`?

Reply:
This is what is used by the Gson serializer, and so I was following that.